### ISSUE TITLE
Feature request: savefig with separated pad_inches in each direction

### ISSUE THREAD

**Issue by scarlehoff** [matplotlib/matplotlib#11764]
This is just a thought I had, it would be desirable to be able to select "pad_inches" separately for the top/bottom/right/left margins.

Just to clarify, I want to be able to create my figures and whatever and only at the time of saving them to disk be able to do:

`figure.savefig(filename,  bbox_inches = 'tight',  pad_inches_top = 0.4, pad_inches_right = 0.3)`

Instead of having to change the shape of the figure I'm working with or creating a new figure with the right dimensions, as the padding work in all directions at the same time. Of course it's not a critical thing but in terms of usability would be a great help.

ps. I'm not sure whether this is a good idea or whether this is the correct place to submit a feature request. Apologies in advance.

**Comment by jklymak** [comment]
Yes this is a good place to submit this idea.  I don’t think it’d be a big deal to pass a four-tuple in pad_inches; I’d be against four new kwargs.   Someone just has to write the PR!

**Comment by jklymak** [comment]
I think if you are going to do a dictionary, you might as well just do the explicit kwargs?

Its kind of funny that `SubplotParams` has the kwarg order lbrt, but the documentation as `lrbt`.  I always get confused about kwargs, but can `__init__` be called w positional arguments?  I don't see anywhere in the codebase where `fig.subplots_adjust` is called with positional arguments, so I somewhat suspect no one was really paying attention to the order of the kwargs.

**Comment by timhoffm** [comment]
> I think if you are going to do a dictionary, you might as well just do the explicit kwargs?
Not quite. The call is comparable complex or lengthy with both versions, but the API stays simpler: Just one arg.

Since the doc is `lrbt`, it's probably an unintended mistake to have the order `SubplotParams(left, bottom, right, top)`. However, even if we don't use that internally, it's now part of the public API and not easy to change.

**Comment by fredrik-1** [comment]
And (matplotlib) basemap use for same reason, left, right, top, bottom. 

I believe that the best API would be to let `pad_inches` be either a float or dict. I don't think that it would be a good idea to have four new kwargs for a functionality that is probably not going to be used that much. A tuple would also be good, less wordy but more difficult to remember.

**Comment by zjpoh** [comment]
I agree that tuple is not the best option here. I also agree that adding one dict is better than four new kwargs for easy of maintenance.

If nobody is working on this yet, I would like to give it a try.

**Comment by jklymak** [comment]
My objection to a dictionary is the same as my objection to fontdict for text.  There is no discoverability for it from the command line of IPython or Jupyter notebook.   Further to fill in a dictionary properly you have to get the keys correct; is it left-pad, leftpad, pad-left, padding-left etc etc etc.  At least with a kwarg you can tab-complete to get the name.  

A tuple on the other hand is easy to get wrong, but there are really only two sensible ways it can be organized so at most you make two guesses.  

I understand why we have fontdict, but I have to google the entries every time I use it and that seems like bad API to me.

**Comment by timhoffm** [comment]
If we use a dict as param it should be described fully in the docstring of the respective param. I know that's not the case for fontdict. You don't get tab-completion from that, but at least you can look it up in the docstring without leaving IPython / the notebook.

To that extent, a dict and a tuple are equally (non-)accessible. The dict has the advantage of being more explicit at the cost of being more verbose. I'd prefer that for rarely used options where there is no unambiguous order as in this case.

Fontdict has another issue in that it's redundant to other kwargs, cf #10293. I think the situations between here and fontdict are not quite comparable. We should not mix this into the discussion here.

**Comment by jklymak** [comment]
Just registering my general disaproval of packing method arguments in dictionaries;  I don't really care about this kwarg in particular, but I think its a klunky API style in general.

**Comment by timhoffm** [comment]
## Short summary of options:

1. `pad_inches` supports tuple `(left, right, bottom, top)` (maybe alternative order, but the one given here seems prevalent within matplotlib - though we are not consistent with ourselves)
2. Separate keywords `pad_inches_left`, `pad_inches_right`, ...
3. `pad_inches` supports dict `{'left'=..., 'right'=..., 'bottom'=..., 'top'=...}`.

Not adding the feature or supporting multiple ways are theoretically possible as well.

Personally, I'm 50/50 on 1. *or* 3. But there were also other opinions. For a detailed discussion of the advantages/disadvantages see above.

I think this is an important API decision, which should be taken deliberately. @tacaswell @efiring Can please some of the senior contributors decide on the API we want to use here.

**Comment by efiring** [comment]
Considerations: discoverability, readability, and code complexity, all viewed with the perspective of how often a feature will be used, and by whom.

Discoverability involves documentation as well as possible tab-completion. Having a large number of kwargs hurts discoverability; the docstring becomes overwhelming, and there is a problem even for tab-completion, given the present ipython default of providing a quasi-gui interface instead of listing everything.

Readability is high-priority, and having the option of packing related options into a single dict kwarg aids readability, compared to an opaque tuple with non-obvious ordering.

The option proposed here, allowing more detail in pad_inches, is to me low-priority; I think its use will be rare.  This puts extra weight on readability (if the option is indeed worth adding at all).

If this is implemented, I think options 1 (tuple) and 2 (more kwargs) is perpetuating bad API; option 3 (allowing a dict) is by far superior.

**Comment by EzraBC** [comment]
Be aware, this is my first attempt at contributing to something.

I propose to modify matplotlib.transforms.BBoxBase.padded to return a Bbox with the points modified by values in a dict as opposed to a single value. If a single value is passed, it can be normalized via the dict.fromkeys method within BboxBase.padded as shown [here](https://github.com/matplotlib/matplotlib/compare/master...EzraBC:master).

**Comment by Roald87** [comment]
Is there something against using a `namedtuple`?

**Comment by NeilGirdhar** [comment]
Instead of `namedtuple`, I would just support any object that exposes the four attributes you want.  For convenience, provide a `dataclass` type.

**Comment by jklymak** [comment]
Just to summarize above, I think the plan is to use a dict if someone cares enough to submit a PR...

**Comment by zjpoh** [comment]
I'll do it.

**Comment by ttshaw1** [comment]
Can we reopen this? @zjpoh is this on your to-do list? I can take it over if not, this would be useful for a graph I'm making right now

**Comment by jklymak** [comment]
You are welcome to put in a PR even if the issue is closed

**Comment by ttshaw1** [comment]
I opted for a [l,r,b,t] tuple because I think that's going to be the most accessible. Additional kwargs and more complex datatypes seem like overkill for most users to me. If I'm understanding the above correctly, `SubplotParams()`'s lbrt order is the black sheep and a lrbt tuple will be reasonably conventional. Can't speak to tab-completion as I don't think it's functional in my Spyder environment

**Comment by timhoffm** [comment]
Semi-OT: Very few users seem to rely on the parameter in SubplotParams(): https://github.com/search?q=%2F%5CbSubplotParams%5C%28%5B%5Cw%5C.%5D%2B%2C%2F+language%3APython+NOT+is%3Afork&type=code

Therefore, it might be an option to make its parameters kwonly. This would remove the inconsistency. It also adds to clarity, and  SubplotParams Is such an advanced use case that we can bear the additional verbosity.

Edit: It's more complicated than that, as this also has to be done in other places like `SubplotParams. update` and `Figure.subplots_adjust` ([usage](https://github.com/search?q=%2F%5C.subplots_adjust%5C%28%5B%5Cw%5C.%5D%2B%2C%2F+language%3APython+NOT+is%3Afork&type=code)). So likely not worth it. I think subplot params are generally less used nowadays since constrained layout is often a good automatic default.

**Comment by timhoffm** [comment]
Summarizing the current APIs we have (partially taken from [above](https://github.com/matplotlib/matplotlib/issues/11764#issuecomment-407485971)):

- `Axes.imshow(extent=)` uses `(left, right, bottom, top)` - while they share the same structure, at least they are positions not pad
- `SubplotParams(left, bottom, right, top)` - these are pads, but in a weird order
- `tight_layout` (and related functionality) - only scalars as separate parameters: `pad`, `h_pad`, `w_pad`.

So no matter what we do with a 4-tuple: It's a different API than tight_layout and it will be inconsistent either with the subplot_params or extent tuples.

---

If we want to expand pad semantics, I'm inclined to go with the CSS margin / padding approach; i.e. accept
- a single value for all cases
- two values (vertical, horizontal)
- four values (top, right, bottom left)

Advantages:
- a widely known convention and thus better memorizable
- convenient as often the two-value version is sufficient (and reduces the need to think too much about the order of the individual sides)
- can easily be retrofitted to any place we can currently have only a scalar pad (including tight_layout)

Disadvantages:
- Inconsistent with the order in subplot_params - but hopefully less used nowadays
- Inconsistent with the order in extent. - we need to clearly document this.

**Comment by ttshaw1** [comment]
I agree it makes sense for imshow, SubplotParams, and tight_layout (and largely anything else where we're specifying a bounding box type of thing) to accept all of
- a single value
- specify the x and y axes
- specify all four bounds

But I want to point out that the left, right, bottom, top order is used in other contexts in matplotlib. Off the top of my head, hist2d's `range` ([[xmin, xmax], [ymin, ymax]]), and imshow's `extent` ([l, r, b, t]), though I'm sure there are a lot of other examples. In the typical "turn data into a pretty png for a powerpoint" tasks that I do, I see l, r, b, t quite a bit and other orders never.

I'd vote for any argument in this bounding box-ish domain to accept 
- a single number or len-1 list-like to specify all
- [x, y] list-like
- [left, right, bottom, top] list-like
wherever we can do that without breaking changes

**Comment by timhoffm** [comment]
> , hist2d's `range` ([[xmin, xmax], [ymin, ymax]]), and imshow's `extent` ([l, r, b, t])

This is not directly comparable. Logically, it’s a lis of ranges. A grouping [[pad_l, pad_r], [pad_top, pad_bottom]] wouldn’t make sense.

It think we need to create a systematic overview of all cases involving „sizing“ in x and y to make an informed decision. `Rectangle` and `Bbox` directly to my mind. Though not all of these will be logically equivalent (see hist2d‘s range) I think it’s valuable to have the broader picture.

**Comment by jklymak** [comment]
I agree that these things are just convention and it would be nice to have one paradigm in Matplotlib, though for back compat reasons I expect this will be hard.  

I find `[t, r, b, l]` super unintuitive, and I'm pretty surprised that is the CSS default.  However, it makes some sense in CSS to start from the top, since web layout is from the top down (and then they went clockwise, I guess).  I'm also unaware of it _ever_ being used in Matplotlib.

However, for Matplotlib, `ax.set_position` is `[l, b, w, h]`.  For `Bbox`: "extents" are `[l, b, r, t]`, and "bounds" are `[l, b, w, h]`.  Given that these are super fundamental to Matplotlib, I'd suggest we keep the same convention and use `[l, b, r, t]`.  I dont' find `Rectangle` as useful an analogy, but it is specified `(xy), width, height` where `(xy)` is "typically" the bottom left corner.  

So I think its pretty universal for Matplotlib (and Matlab before it) to start "left, bottom,..." I can see where the other convention comes from, but I think it's just a different paradigm.

**Comment by timhoffm** [comment]
Given that there is no canonical order of the values and that this is and advanced feature used for fine-tuning, I vote to be defensive and only accept a dict for specifying per-side pads (option 3 from above).

Advantage:
- makes per-side pads possilbe
- unambiguous
- we can always add tuple support later once/if we come to an agreement this is worth it and we have a consistent memorizable API

Disadvantage:
- more typing, in particular if you want vertical pads and horizontal pads (i.e. the dict doesn't have an equivalent for the 2-tuple variant)
